### PR TITLE
Add Ipopt

### PR DESCRIPTION
**Pull request by jschueller**
- Add ipopt solver (~ same internal interface as bonmin)
- Remove OPT++ solver (too much freeze/nan/crashes/exit, ..., see mathieu's report)

### CODE REVIEW

**Comment by jschueller** [comment]
ping @sofianehaddad @regislebrun

**Comment by regislebrun** [comment]
I saw it, but you know, from time to time I take some holidays...

**Comment by jschueller** [comment]
yes ;) I should do too

File: lib/src/Base/Optim/IpoptProblem.cxx [hunk 1]
```diff
@@ -0,0 +1,575 @@
...
+                               ::Ipopt::IpoptCalculatedQuantities* ip_cq)
+{
+  // Convert x to OT::Point
+  std::copy(x, x + n, optimalPoint_.begin());
+  if (optimProblem_.isMinimization())
+    optimalValue_[0] = obj_value;
+  else
+    optimalValue_[0] = -obj_value;
+}
+// void IpoptProblem::finalize_solution( TMINLP::SolverReturn status,
```
Comment (regislebrun) [inline lib/src/Base/Optim/IpoptProblem.cxx]: Dead code?

File: lib/src/Base/Optim/IpoptProblem.cxx [hunk 2]
```diff
@@ -0,0 +1,575 @@
...
+ */
+#include "openturns/IpoptProblem.hxx"
+#include "openturns/SpecFunc.hxx"
+
+using namespace Ipopt;
+
+BEGIN_NAMESPACE_OPENTURNS
+
+/** Constructor with parameters */
+IpoptProblem::IpoptProblem( const OptimizationProblem & optimProblem,
```
Comment (regislebrun) [inline lib/src/Base/Optim/IpoptProblem.cxx]: why no const here? Side effect on optimProblem?

File: lib/src/Base/Optim/IpoptProblem.cxx [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+                              const Point & startingPoint,
+                              const UnsignedInteger maximumEvaluationNumber)
```
Comment (regislebrun) [inline lib/src/Base/Optim/IpoptProblem.cxx]: const here

File: lib/src/Base/Optim/IpoptProblem.cxx [hunk 4]
```diff
@@ -0,0 +1,575 @@
...
+  {
+    // The definition of the Lagrangian used in Bonmin is available in Ipopt online documentation
+    // at https://coin-or.github.io/Ipopt/classIpopt_1_1TNLP.html#a26b9145267e2574c53acc284fef1c354
+
+    // Convert x to OT::Point
+    Point xPoint(n);
+    std::copy(x, x + n, xPoint.begin());
+
+    // Compute objective hessian
+    SymmetricMatrix objectiveHessian(obj_factor * optimProblem_.getObjective().hessian(xPoint).getSheet(0));
```
Comment (regislebrun) [inline lib/src/Base/Optim/IpoptProblem.cxx]: const

File: lib/src/Base/Optim/IpoptProblem.cxx [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+
+    // Compute constraints hessian
+    int k = 0;
+    SymmetricMatrix constraintsHessian(n, 0.0);
+
+    if (optimProblem_.hasEqualityConstraint())
+    {
+      SymmetricTensor equalityConstraintHessian(optimProblem_.getEqualityConstraint().hessian(xPoint));
```
Comment (regislebrun) [inline lib/src/Base/Optim/IpoptProblem.cxx]: const

File: lib/src/Base/Optim/IpoptProblem.cxx [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+      for (UnsignedInteger i = 0; i < nbEqualityConstraints; ++i)
+      {
+        constraintsHessian = constraintsHessian + lambda[k] * equalityConstraintHessian.getSheet(i);
+        ++k;
+      }
+    }
+
+    if (optimProblem_.hasInequalityConstraint())
+    {
+      SymmetricTensor inequalityConstraintHessian(optimProblem_.getInequalityConstraint().hessian(xPoint));
```
Comment (regislebrun) [inline lib/src/Base/Optim/IpoptProblem.cxx]: const

File: lib/src/Base/Optim/IpoptProblem.cxx [hunk 7]
```diff
@@ -0,0 +1,575 @@
...
+      for (UnsignedInteger i = 0; i < nbInequalityConstraints; ++i)
+      {
+        constraintsHessian = constraintsHessian + lambda[k] * inequalityConstraintHessian.getSheet(i);
+        ++k;
+      }
+    }
+
+    // Filling 'values' array
+    SymmetricMatrix lagrangianHessian(objectiveHessian + constraintsHessian);
```
Comment (regislebrun) [inline lib/src/Base/Optim/IpoptProblem.cxx]: const

File: lib/src/Base/Optim/openturns/IpoptProblem.hxx [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+                    int* jCol,
+                    double* values);
+
+  virtual void finalize_solution(::Ipopt::SolverReturn status, ::Ipopt::Index n,
+                               const ::Ipopt::Number* x, const ::Ipopt::Number* z_L,
+                               const ::Ipopt::Number* z_U, ::Ipopt::Index m, const ::Ipopt::Number* g,
+                               const ::Ipopt::Number* lambda, ::Ipopt::Number obj_value,
+                               const ::Ipopt::IpoptData* ip_data,
+                               ::Ipopt::IpoptCalculatedQuantities* ip_cq);
+//   void finalize_solution( ::Bonmin::TMINLP::SolverReturn status,
```
Comment (regislebrun) [inline lib/src/Base/Optim/openturns/IpoptProblem.hxx]: Dead code?

**Pull request merged**